### ISSUE TITLE
DPDK for Xilinx/AMD Alveo U50 and U280 100GbE

### ISSUE THREAD

**Issue by AdamI75** [missinglinkelectronics/dpdk-stable#4]
I am using an Alveo U50 and Alveo U280, which is not what you tested against and is 100GbE . I am able to compile and build the drivers using meson and ninja. I have configured the Alveo U50 and U280 with Corundum and when I load the kernel I can see the Corundum NIC interfaces, ping them and send data via iperf3. However, when I run the DPDK command: "./dpdk-devbind.py --status" from the MLE DPDK repo "usertools" directory, I do not see the Corundum NICs whether I load the kernel or not. How should I enable the PMD? It says it was tested with a slightly older version of Corundum and only for 10GbE. Therefore, does this mean that it won't work for the Alveo U50 and U280 as they are 100GbE? I am not sure. I am still reading up and will continue to try on my end. I would like to create a test where I can produce data from one Alveo configured with Corundum on my one server and then send that to another Alveo configured with Corundum on another server using iperf3 and then load the mqnic kernel driver and compare with your DPDK MLE mqnic driver. I am concerned that the script is not seeing the Corundum configured Alveos. It may need further porting. What do you think? Thanks!

**Comment by AdamI75** [comment]
It looks like dpdk-devbind.py needs to updated to reflect the mqnic drivers. Do you have some examples or documentation on how to do this? Maybe you can email me the file that you use: aisaacson@sarao.ac.za, thanks? I will continue to play around for now until I get confirmation.

**Comment by AdamI75** [comment]
Yes, that is the case. The Alveo is seen as a memory controller and not a network device, so I changed the class to 5 in the python script and now I see it. I will keep you posted.

**Comment by AdamI75** [comment]
I can now bind the Alveo U50 to the vfio-pci driver - see screen attached. However, as soon as I do this then I no longer see the NIC interface using ifconfig. I am also unable to use dpdk-testpmd or the layer 2 forwarding app - see second screen shot with error. If I run these apps then it fails to create the interface. However, if I use the normal mqnic driver (non-DPDK) then the application will run. Am I binding this to the right driver? I am not really using any virtual machines. Please can you provide a step by step account of how you binded your driver and got the dpdk-testpmd app to run, thanks? Is there a way to still see the NIC interface with ifconfig once the driver has been bound to the DPDK compatible driver?

Status update: This actually makes sense that I can't see the NIC via ifconfig as I am no longer using the kernel driver and so I need to use the DPDK API calls to access the PHY and other parameters. I am just not sure why testpmd and l2fwd is not working with this driver and why it is not creating the interface? Maybe you can give me your explanation or point me in the right direction, thanks? I am stumped for now :). A few steps on how you did your testing would be great, thanks :). 

![DPDK_compatible_driver_17_Aug_2023](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/f6e4e4b7-f768-44ab-871a-a93097705a70)

![testpmd_fail_17_August_2023](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/159e60ce-4cf1-4b68-b88c-ca9977ee97b6)

**Comment by joft-mle** [comment]
Hi @AdamI75 ,

1. Yes, this DPDK driver for mqnic does not support running the kernel mode driver and the DPDK driver in parallel.
2. That the class code needs to be changed - is strange. Currently we have no idea why the FPGA design would come up with such a "wrong" class code. Some IP Core/TCL code modification?
3. As per the second screen shot in your posting above, the loaded mqnic FPGA design is a built of Git commit https://github.com/corundum/corundum/commit/ed4a26e2, which is pretty recent (2023/07/13). **This DPDK driver has however been used and tested specifically with Git commit https://github.com/corundum/corundum/commit/0965c77b8e787f56bed84d539821fec448cd24a3 (2021/10/02).** Since then the register interface has been changed/updated here and there. In your case the message "TX queue block not found" almost at the end of the output says that a _compatible_ TX Queue Manager block could not be found in the FPGA design.

Independent of the above, of course we should not run into a `Segmentation fault`, upon not finding a compatible register interface. @pmo73 , we should fix this at the next opportunity.

**Comment by AdamI75** [comment]
@joft-mle thanks very much for this useful info. I am busy compiling with your version for now. I will leave the U280 with the latest Corundum version and just check the U50 with the version you used. I hope this does the trick - sounds like it might. I am still compiling using Vivado 2023.1 and I am not sure what version of Vivado you are using unless you specified it somewhere that I missed?

I will comment out the class addition and report back on what I find. Maybe it is to do with the Vivado version too?

I know you use dpdk-testpmd to verify and l2fwd. Do you mind specifying the exact commands with EAL arguments for testing, thanks?

**Comment by AdamI75** [comment]
@joft-mle Okay, now we seem to be getting somewhere. It now fails at "FW ID block not found" - see screen capture and I commented out the class 5 line and it still sees the Alveo, so I think there have been some register changes to the new version of Corundum as you said.  I suspect that if I use your version of Vivado to compile the design then it will find all the required blocks. I am wondering if I should edit the latest Corundum project source rather. The version of the Corundum repo you are using throws errors for the kernel driver build session.

At least this makes more sense now.

![FW_ID_Block_not_found_18_August_2023](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/0c0a79e5-669f-4767-8ba0-02dd26da6bab)

**Comment by joft-mle** [comment]
@AdamI75 , no, unfortunately it really looks like the mqnic DPDK driver is failing even earlier this time. The Firmware ID block is the very first set of registers any mqnic driver expects to see - at offset 0 of the first PCI BAR. The type code is 0xffffffff and your output shows `type 0x00000000` instead. The version of this first and only register block `(v 0.0.0.1)` also looks odd. I think it should be `(v 0.0.1.0)`, if I'm not mistaken.

Thus I can only assume that the FPGA design has not been loaded properly? Does the situation change after reprogramming the FPGA card and even reboot the host machine?

**Comment by alexforencich** [comment]
"FW ID block not found"...are you running Corundum or the Alveo image?

**Comment by alexforencich** [comment]
Ah, the linked commit from MLE is for a very, very old version of Corundum.  The register space has seen a rather complete overhaul since then.  Until the DPDK PMD is updated, it cannot be used with the latest version of Corundum.  Also, I think their PMD code is missing some checks that would prevent it from even trying to talk to a recent version of Corundum that uses the new layout.

**Comment by AdamI75** [comment]
@joft-mle and Alex. I did the following:
1) I did a cold boot, reloaded the U50 MLE Corundum version (https://github.com/corundum/corundum/commit/0965c77b8e787f56bed84d539821fec448cd24a3) (2021/10/02). 
2) I did a warm reboot
3) I checked with lspci that Corundum PCIe settings were present: lspci -d 1234: -v. It picked it up.
4) I checked that the Alveo U280 still had the Xilinx shell loaded: lspci -d 10ee: -v. It picked it up.
5) I then used the DPDK repo MLE version (22.11-add-corundum-mqnic-pmd branch latest githash) and under "dpdk-stable/usertools" I first ran "./dpdk-hugepages.py -s" to see nothing was setup, then I ran "./dpdk-hugepages.py --setup 2G" and then "./dpdk-hugepages.py -s" to see the hugepages were setup.
6) I then bound the DPDK driver to my Alveo U50 with MLE version Corundum image to c2:00.0, which is my Alveo U50 location. I used "sudo ./dpdk-devbind.py -b vfio-pci c2:00.0" to bind and then "./dpdk-devbind.py -s" to check the status. It looks fine.
7) I then tried the helloworld app just to see if there were issues. I changed directory to the "build/examples" folder and ran "./dpdk-helloworld -a c2:00.0". Unfortunately, I still get the "FW ID block not found issue", so I am certain that I am doing this correctly. The firmware is loading correctly. I am using Vivado 2023.1 to compile the MLE Corundum image and I wanted to know what Vivado version you are trying, so that I can try this too - just to see if it is an issue or not.
8) I tried to run Alex's makefile for his drivers in the MLE version of Corundum, but it does not build due to various errors. I think this is what Alex is alluding to in his above comments with register map updates etc.
9) I then tried to load Alex's Kernel module from the latest Corundum checkout, but it does not detect the NIC interface probably as the register maps are different.
10) I then reporgrammed the U280 and U50 with Alex's latest Corundum repo and "sudo insmod <path to mqnic driver>/mqnic.ko and everything ran correctly with the kernel driver. I can use iperf3, ping and do all my testing.

I am busy documenting the ease of building drivers for Corundum as we decide on a suitable framework for our SmartNIC. It seems to me that the DPDK still needs some fine-tuning before others can use it. I am not sure how involved MLE is with Alex or what your priorities are like. It just seems that a bit more effort is required before users can use the DPDK drivers as it stands? I have only spent a few days on this, so I am a newbie :). My report will highlight this.

This is my wish list for DPDK - not sure whether you can make it come true or not :). I would like the following:

1) A nice step by step guide on your install, build and testing with your release would be very useful. I know this is documented in the DPDK documentation, but not all steps are covered in detail and some assume some sort of prior knowledge. Of course, infrastructure plays a role, but it is useful to see how this is handled on one system, step by step.

2) Please port to the latest Corundum version. I want to be able to switch between the kernel driver and DPDK for testing.  There should only be one version with one register map. I am not sure what your objectives and needs are right now, but I may be able to assist if work needs to be done.

3) I need to be able to compare the kernel driver with DPDK driver. This means that I probably need to use iperf3 for the kernel driver and pktgen for the DPDK driver - correct me if I am wrong. These drivers can only be used mutually exclusive from one another, from what I understand. Some PMDs can be used with the kernel driver e.g. Mellanox, but mqnic cannot. This is fine, but it needs to be documented in a better way with suggestions on how to compare the kernel testing with DPDK testing.

4) If possible, please can you test your DPDK framework on an Alveo U50 or U280 instead of just the Xilinx ZCU106 (I think you used this one). Maybe there are hardware differences that your driver is not catering for. The Alveos run shells and the ZCU106 may not? I think it is a development card with PCIe capability where the Alveos are PCIe cards.

Okay, I think I have enough info for my report for now. I will continue testing with Alex's kernel module, which works fine for now. Thanks for the support. I really appreciate it. Best of luck!

**Comment by joft-mle** [comment]
Hi @AdamI75 ,

> 7. I then tried the helloworld app just to see if there were issues. I changed directory to the "build/examples" folder and ran "./dpdk-helloworld -a c2:00.0". Unfortunately, I still get the "FW ID block not found issue", so I am certain that I am doing this correctly. The firmware is loading correctly. I am using Vivado 2023.1 to compile the MLE Corundum image and I wanted to know what Vivado version you are trying, so that I can try this too - just to see if it is an issue or not.

I'm very sorry, I just realized today that the commit message of https://github.com/missinglinkelectronics/dpdk-stable/commit/56ad0e5b53edf4678fa40bbccacfda6e82fcd597 "simply" :-( points to the wrong Corundum Git state. Really embarrassing :-( .

Instead of https://github.com/corundum/corundum/commit/0965c77b8e787f56bed84d539821fec448cd24a3, this DPDK driver has been developed and tested, at least in its final state, with:

**=> Corundum Git state https://github.com/corundum/corundum/commit/56fe10f27d9b42f1ff9abe4d735b113008e4be9d (2022/09/22)**

We will come up with a separate commit, adding a README file, which specially points to that correct version. While we are at it, we will include at least a minimal step-by-step guide regarding testing.

And, another sorry: on Friday I did not answer your question about which Vivado version we were using: **Vivado 2021.2**.

**Comment by AdamI75** [comment]
@joft-mle No worries, I learnt a lot from the whole process :). Thanks for the update. I will try this - one more time - and get back to you. I will still continue with Vivado 2023.1, as it shouldn't make a huge difference from what I have seen. Please send me the test document when you can, thanks - I really do appreciate the effort.

**Comment by AdamI75** [comment]
@joft-mle and @alexforencich: Okay, I tested the Corundum Git state https://github.com/corundum/corundum/commit/56fe10f27d9b42f1ff9abe4d735b113008e4be9d (2022/09/22) that was used for DPDK testing and I can say the following:

1) The hello world and l2fwd application runs, so this is great news. However, the link status shows 10Gbps and the  MAC  address seems different from the kernel driver. See screen capture 
![DPDK_l2fwd_23_August_2023](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/288759b3-bed3-4bfb-9ad6-862fc78ac1ea)


2) I still cannot build the standard kernel driver that comes with mqnic - there are errors (see screen captures). 
![Errors_with_kernel_building_23_August_2023](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/4f9b095b-7592-4262-874d-9bda7541d84f)

3) The dpdk-stable version (githash 56ad0e5b53) that I used from MLE, still requires me to edit dpdk-devbind.py to include a class 5 in order to see the Alveos. See screen capture - 
![DPDK_python_bin_script](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/dc88341f-a4de-4a84-aee7-d95f8a1294cb)

My suggestion is that you rebuild and test using the latest corundum githash repo as Alex requested you use and it would help to have a step by step test document, so that I can verify your driver and understand how to set the link to 100Gbps. Good effort though and I really appreciate the support.

**Comment by alexforencich** [comment]
Link status is bunk as Corundum doesn't report the link speed anywhere right now, the link speed is effectively fixed in the FPGA design - a 100G design will only run at 100G, etc.  Currently the 25G design can be reconfigured dynamically to run at 10G, but this is still little more than a prototype capability and it isn't yet fully integrated.  This may change later on, but it's not a high priority.

MAC addresses are also bunk.  The kernel driver has a bunch of board-specific code for handling that, but the plan is to delete that and have a soft core on the FPGA handle that kind of board-specific interfacing.  This is also rather low priority at the moment, but once it's implemented then both the kernel driver and DPDK driver will have a board-independent way to read the default MAC addresses, interface with the optical transceivers, and reconfigure the transceivers for different link speeds.

Those bugs in the kernel driver are likely fixed in the current version of Corundum.  I don't back-port changes like that, but I think it's just a couple of ifdefs to handle changes in the kernel API.

Strange that the class code is wrong.  It's possible that this was adjusted in the TCL scripts at some point.  What does lspci show for the class code?  What does the PCIe config GUI in Vivado show as the configured class code?  Does the current git master version behave differently?

**Comment by AdamI75** [comment]
@alexforencich Thanks for the info on the MAC and link speed. That means it has probably been correctly configured. Good to know. 

You are right that the latest Corundum commit kernel driver builds without error. I have tested this, but I am not too worried about this right now. I know which MLE Corundum build to test and when @joft-mle sends me a step-by-step test document then I will verify the DPDK drivers using their test on the Alveo U50 and U280. By the way, the 81:00.0 is the U280 and the c2:00.0 is the U50, FYI.

Yes, I have done an "lspci -d 1234: -v", the screen capture is below.  ![lspci_Alveo_U50_U280](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/7477f1c8-0bc2-440c-9713-01394deec55d)

You will see they refer to the U50 and U280 as "memory controllers". If you look at the PCI SIG class system, this should be class 5 and not class 2 - see page 8 https://pcisig.com/sites/default/files/files/PCI_Code-ID_r_1_11__v24_Jan_2019.pdf. I think you have not encountered this issue before because you have not tested using the Alveos.  Anyway, we have a fix for this, if necessary. It is only when I run ./dpdk-devbind.py from the DPDK MLE repo tested githash. I didn't need to make modifications for your repo using your kernel driver. Yes, the Vivado GUI for the PCIe IP shows class 5 - see screen capture. 
![Vivado_memory_class_5](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/16c06939-050a-4f6a-a58c-c420ad73ffe0).

Yes, your latest version of Corundum (githash ed4a26e) displays memory controller as class 5, so I think the dpdk-devbind.py file will need to be updated eventually (unless I just change the Vivado IP to indicate class 2), but as you said the Alveo is not meant for networking. It is just that we have these cards available to use for now and will probably order a more suitable card at some point.

**Comment by joft-mle** [comment]
@AdamI75 , a few more details regarding your [second to last post](https://github.com/missinglinkelectronics/dpdk-stable/issues/4#issuecomment-1689496618)'s items:

1) The DPDK driver currently simply reports 10G in any case, see source code, function `mqnic_dev_info_get()`. So you'll get the meta information "10 Gbps" for any mqnic design. Also, the topic of MAC addresses is kept rather simple in the driver - as can be seen in function `eth_mqnic_dev_init()`, the MAC address is always a randomly generated one. The upstream kernel driver, for most supported boards, tries to read (a) dedicate address(es) from a board's EEPROM; otherwise the kernel driver also uses randomly generated addresses. However this functionality (try read from EEPROM) has not been integrated into the DPDK driver yet. And as @alexforencich  pointed out, the MAC address situation will eventually change significantly in the future anyway. As with the Linux kernel, I would be surprised, if a DPDK application doesn't have the possibility to set a MAC address via some API call. So in practice there should be a easy workaround possible for this MAC address shortcoming.

2) Next to backporting these few changes to the kernel driver, as explained by @alexforencich , you could also simply use an older kernel. I guess, you are using a rather current kernel version - v6.1 ? That older state of the driver in https://github.com/corundum/corundum/commit/56fe10f27d9b42f1ff9abe4d735b113008e4be9d builds cleanly against v5.4 (sure) and possibly for v5.10 (unsure).


Regarding updating to the DPDK driver to current Corundum: Unfortunately we currently do not have any plans to start that effort soon. Sure, the likelihood that it will happen in the future is not low, but it is not on the horizon at the moment. So for the moment, we do provide this DPDK driver as is and also regard it as a potential starting point for somebody else to pick up.

Regarding the strange class code 5 topic: I'm not really familiar with the Alveo cards, but isn't it simply a question of setting the right values/properties on the PCIe IP Core instance in Vivado - using TCL?! Perhaps the Vivado machinery is somehow re-setting this class code value to 5 after it was initially set to 2 via @alexforencich 's TCL scripts? Wouldn't be the first time that Vivado silently overwrites IP core instance properties.
Both, the TCL script in the current upstream (https://github.com/corundum/corundum/blob/ed4a26e2cbc0a429c45d5cd5ddf1177f86838914/fpga/mqnic/AU50/fpga_100g/ip/pcie4c_uscale_plus_0.tcl#L15) and the older Git state (https://github.com/corundum/corundum/blob/56fe10f27d9b42f1ff9abe4d735b113008e4be9d/fpga/mqnic/AU50/fpga_100g/ip/pcie4c_uscale_plus_0.tcl#L15) clearly - at least try to - set it to 2.

**Comment by alexforencich** [comment]
Setting the class code via TCL can be a bit screwy as there are basically two ways to set the code - one is to plug it in directly, the other is to use the "lookup assistant" thing.  Looks like the TCL scripts are probably screwed up, I'll take a look at that and see if I can figure out what the proper method is.  

There also seems to be a bug somewhere in terms of the subsystem device ID, at least on ultrascale devices.  For some reason the subsystem vendor ID is reporting the same value as the vendor ID, despite the Vivado UI showing the correct value.  Not sure if this is a Vivado bug (generates the bitstream with the wrong subsystem ID) or a silicon bug (bitstream has the correct value, but the wrong value is exposed in config space).

IMO, this isn't really a major issue as the kernel driver ignores the class code.  Not sure why DPDK cares about it to such an extent; the vendor and device IDs alone should be sufficient to attach a driver.  Class code is mainly for when there is a generic driver available, such as for NVMe devices, in which case the vendor and device IDs should be ignored.

**Comment by alexforencich** [comment]
Looking at your screenshot; the actual class code field is off screen to the right, and the "use lookup assistant" box is NOT checked.  So that "05" should not be used as part of the class code, instead the class code in the box on the right should be used.  Seems like maybe Vivado introduced a bug at some point relating to how the class code is handled in the PCIe core, as it looks fine to be on 2021.2, but you're using 2023.1.  I'll look in to installing 2023.1 and see if I can replicate the issue.

**Comment by AdamI75** [comment]
@joft-mle and @alexforencich thanks for the useful inputs and responses. I will document this in my analysis report. The PCI class driver issue is most likely a Vivado version issue knowing Xilinx. I don't have Vivado 2021.2 and to be honest I think Vivado 2023.1 would be better as they have built in HBM simulation support - finally. I haven't explored this in depth. My focus is to evaluate what exists, how easy it is to use and if there is any work yet to be done covering Corundum FPGA, Kernel drivers and the DPDK drivers. It is a small issue and can be rectified or fixed later on. If you can replicate using Vivado 2023.1 that would be useful, thanks. I think what would be more useful to me at this stage are the steps needed to test the DPDK. Thanks for the support.

**Comment by pmo73** [comment]
Hello @AdamI75 , i updated all three branches for DPDK 20.11, DPDK 21.11 and DPDK 22.11 and added a README file inside "drivers/net/mqnic" with the information, how i set up the device and tested the driver.

**Comment by AdamI75** [comment]
@pmo73 thank you very much. I have reviewed the DPDK 22.11 branch README file and it would be very useful to see the full output using l2fwd and testpmd. Would it be possible to update those screen captures or add multiple important screen shots with important info for guidance, thanks? I am happy if you just update the screen shots and add them here in this issue for now for more detail and debugging.

**Comment by pmo73** [comment]
The log is too long for a README file, but i can provide it to you as a separate file here inside the issue.
[2023_02_06_dpdk_l2fwd.log](https://github.com/missinglinkelectronics/dpdk-stable/files/12462330/2023_02_06_dpdk_l2fwd.log)

**Comment by AdamI75** [comment]
@pmo73 This is perfect, thank you very much.

**Comment by AdamI75** [comment]
@joft-mle , @pmo73 and @alexforencich thanks for the great support. My company have decided to utilise Corundum for our SmartNIC framework. We will eventually focus on the DPDK driver and what I wanted to know what is still outstanding work on the DPDK driver before it can be integrated into Corundum? Is it also possible to make the DPDK driver work with the kernel driver, so that the two don't need to be mutually exclusive? The idea being that the SmartNIC would look like a NIC to the OS whether it uses the kernel driver or DPDK driver i.e. the user could run "ifconfig" using any of the drivers and it will be seen as NIC interface to the OS? I think it is possible with the Mellanox DPDK driver to do this, but I might be wrong. 

Therefore, from my side, I see the following needs to be done before Corundum can integrate with your driver and this is work that my company could do:

1)  It needs to be ported to work with the latest version of Corundum
2) The DPDK repo needs to be integrated properly with the Corundum repo. This means creating an official Corundum DPDK repo, which can be called as a submodule in the original Corundum repo. Just a suggestion.
3) I will investigate the issue with the class 5 vs class 2 when it comes to the IP core and tcl scripts in Vivado 2023.1 and fix this.
4) Sort out the link speed reporting and MAC address reporting
5) add proper documentation and tutorials for testing.

Anything else from your side that needs to be added or considered? Perhaps you have other ideas? This would be future work from our side once we want to optimise our SmartNIC design

**Comment by pmo73** [comment]
Hello @AdamI75 , yes it should be possible to use the DPDK driver with the kernel driver. The Mellanox driver is one of the few, as far as I know, that allows this. I think you are actually looking for a bifurcated driver (https://doc.dpdk.org/guides/howto/flow_bifurcation.html) that uses the kernel driver. I can't tell you how much you need to modify the kernel driver and the corundum design for this to work.

**Comment by AdamI75** [comment]
@pmo73 Thanks, this is useful to know.

**Comment by alexforencich** [comment]
My understanding is that the Mellanox DPDK PMD actually uses the IB packet APIs.  So, that actually might be quite an interesting approach for Corundum as well for a couple of reasons.  First, if everything is done via APIs to the kernel driver, then no register access would be necessary from the DPDK PMD, and hence there should be no versioning issues with the DPDK PMD.  Second, the long-term plan is to implement RDMA, so if we add some IB APIs now we'll be able to build on top of that later on.

**Comment by alexforencich** [comment]
Also @AdamI75 if you guys want to do a quick zoom call or something so we're all on the same page in terms of what needs to be done, I would certainly be open to that.  We can coordinate that on the Corundum zulip at https://corundum.zulipchat.com/, or via email.

**Comment by alexforencich** [comment]
On the PCIe class code front; when Vivado generates the PCIe IP core, it should generate a file called `pcie4_uscale_plus_0.v` which is the top-level file for the core.  In this file, one of the parameters is `PF0_CLASS_CODE`.  What value does that parameter have in your case?

**Comment by AdamI75** [comment]
@alexforencich I see the pcie4_uscale_plus_0.xci file has the value "5" and "0x58000" associated with the PF0_CLASS_CODE if I do a grep through the files. I do not see "2" - see below for a screen shot.

![class_code_memory_controller_grep](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/f004de73-4b39-4b6d-838d-f1189046e5bd)


Thanks for the info on the DPDK PMD and yes, a catch up would be a great idea. We are meeting tomorrow, internally, so I will discuss this with the team.

**Comment by alexforencich** [comment]
Interesting.  Looks like the issue is probably just the TCL script not setting the configuration properly.  Try adding `CONFIG.PF0_Use_Class_Code_Lookup_Assistant {false}` to the list of settings in the PCIe core TCL script and let me know if that helps at all.  I may just need to go add that everywhere.

**Comment by AdamI75** [comment]
@alexforencich thanks for the guidance. I am actually giving a Corundum demo to the team today, but will try this later this afternoon after our internal meeting and get back to you. 

update: Okay, I tested this below. I added in the CONFIG.PF0_Use_Class_Code_Lookup_Assistant {false} to the pcie IP tcl file as follows:

![tcl_PCIe_file_mod](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/4ae1dbc2-c33a-4840-8d74-baac6227661c)

I then saved it and ran the makefile to build the Corundum 100GbE AU50 project. I noticed that it gave this warning regarding the PCIe IP parameter assignment and not being able to overwrite the 0x58000 value, but I am not sure how critical this is.

![screen_shot_PF0_CLASS_CODE_disabled_param_change_ignored](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/1463bfcf-5123-4452-8430-28a19d44276b)

The Lookup assistant was definitely disabled when I opened it up in Vivado. I then used the dpdk-devbind.py from the DPDK MLE repo utility as is and I still could not see the card or BDF number. However, when I add the class 5 code snippets as mentioned above for the memory controller in dpdk-devbind.py then I am able to see the BDF under "Other Network Devices" - see below:

![devbind_screen_capture](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/73c6f7be-9b85-4923-a6ae-db97675ecb2f).

These are the devbind.py code changes I made to read back the BDF of the Alveo U50 - see below:

![mqnic_class](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/0a8c0eec-e6f2-4cfe-9863-e46cc9c95dcf)

![mqnic_network_devices](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/cf4c121c-11c3-4a37-90cb-223edbe94cd3)

This really is not critical and it works. It does make sense that a memory controller should be class 5 and not class 2 based on the PCI SIG documentation for PCIe classes.

This is what I get when I use lspci to show you that it is memory controller and not a network device hardware:

![lspci memory controller](https://github.com/missinglinkelectronics/dpdk-stable/assets/17824112/089a9e91-d1cd-42cc-9d00-41a3cab88900)

**Comment by daviderror** [comment]
@joft-mle 
@alexforencich 

Hi! Has the 56fe10f-fixed corundum been tested on a NetFPGA_SUME board with 4 SFP+ ports? When transferring packets from the second part to the third using the dpdk-l2fwd tool, I observe packet loss, and there is no loss when transferring traffic from the first port to the second:(
